File: leetcode/medium/5-longestPalindromicSubstring.js
// Given a string s, return the longest palindromic substring in s.

// Example 1:

// Input: s = "babad"
// Output: "bab"
// Note: "aba" is also a valid answer.

/**
 * @param {string} s
 * @return {string}
 */
  
var longestPalindrome = function(s) {
  let longestLen = 0;
  let longestPal = "";
  
  for (let i = 0; i < s.length; i++) {
    // odd length palindromes
    let l = i;
    let r = i;
    
    while (l >= 0 && r < s.length && s[l] == s[r]) {
      if ((r - l + 1) > longestLen) {
        longestLen = r - l + 1;
        longestPal = s.substring(l, r + 1);
      }
        l--;
        r++;
    }
    
    //even length palindromes
    l = i;
    r = i + 1;
    
    while (l >= 0 && r < s.length && s[l] == s[r]) {
      if ((r - l + 1) > longestLen) {
        longestLen = r - l + 1;
        longestPal = s.substring(l, r + 1);
      }
        l--;
        r++;
    }
    
  }
  return longestPal;
}

File: leetcode/medium/5-longestPalindromicSubstring.java
class Solution {
    public String longestPalindrome(String s) {
        int longest = 0;
        String output = "";
        
        for (int i = 0; i < s.length(); i++) {
            int j = i;
            int k = i;

            while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
                if (k - j + 1 > longest) {
                    longest = k - j + 1;
                    output = s.substring(j, k + 1);
                }
                j--;
                k++;
            }
            
            j = i;
            k = i + 1;
            
            while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
                if (k - j + 1 > longest) {
                    longest = k - j + 1;
                    output = s.substring(j, k + 1);
                }
                j--;
                k++;
            }
        }
        return output;
    }
}
